### PR TITLE
JAMES-1825 use a MessageUid strong type instead of long in the whole …

### PR DESCRIPTION
**Pull request by mbaechler**
…code base

With a lot of help from @rouazana !

### CODE REVIEW

File: mailbox/api/src/main/java/org/apache/james/mailbox/MessageUid.java [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+        this.uid = uid;
+    }
+    
+    public MessageRange toRange() {
+        return MessageRange.one(this);
+    }
+    
+    @Override
+    public int compareTo(MessageUid o) {
+        return Long.valueOf(uid).compareTo(Long.valueOf(o.uid));
```
Comment (rouazana) [inline mailbox/api/src/main/java/org/apache/james/mailbox/MessageUid.java]: Long.compare?

File: mailbox/api/src/main/java/org/apache/james/mailbox/model/MessageRange.java [hunk 2]
```diff
@@ -208,35 +196,21 @@ public String toString() {
...
+    private final class RangeIterator implements Iterator<MessageUid> {
 
         private final long to;
         private long current;
         
-        public RangeIterator(long from, long to) {
-            this.to = to;
-            this.current = from;
+        public RangeIterator(MessageUid from, MessageUid to) {
+            this.to = to.asLong();
```
Comment (rouazana) [inline mailbox/api/src/main/java/org/apache/james/mailbox/model/MessageRange.java]: Why not keeping MessageUid in this class?

File: mailbox/api/src/main/java/org/apache/james/mailbox/MessageUid.java (hunk 1 above)
Comment (rouazana) [inline mailbox/api/src/main/java/org/apache/james/mailbox/MessageUid.java]: or Long.valueOf(uid).compareTo(o.uid) in Java 6...

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java [hunk 3]
```diff
@@ -531,26 +534,26 @@ private Where selectAll(CassandraId mailboxId, FetchType fetchType) {
...
+            .and(lte(IMAP_UID, to.asLong()));
     }
 
-    private Where selectMessage(CassandraId mailboxId, long uid, FetchType fetchType) {
+    private Where selectMessage(CassandraId mailboxId, MessageUid uid, FetchType fetchType) {
         return select(retrieveFields(fetchType))
             .from(TABLE_NAME)
             .where(eq(MAILBOX_ID, mailboxId.asUuid()))
-            .and(eq(IMAP_UID, uid));
+            .and(eq(IMAP_UID, uid == null ? null : uid.asLong()));
```
Comment (rouazana) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java]: ternary spotted ^^

Comment (rouazana) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java]: and anyway it is strange that a selectMessage method can take a nullable

File: mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/CassandraUidProviderTest.java [hunk 4]
```diff
@@ -94,7 +94,8 @@ public void nextUidShouldGenerateUniqueValuesWhenParallelCalls() throws Exceptio
         int nbEntries = 100;
         long nbValues = LongStream.range(0, nbEntries)
             .parallel()
-            .map(Throwing.longUnaryOperator(x -> uidProvider.nextUid(null, mailbox)))
+            .mapToObj(x -> x)
```
Comment (rouazana) [inline mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/CassandraUidProviderTest.java]: Function.identity?

File: mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/json/IndexableMessage.java [hunk 5]
```diff
@@ -149,7 +150,7 @@ private void generateText() {
 
     @JsonProperty(JsonMessageConstants.ID)
     public Long getId() {
-        return id;
+        return id == null ? 0 : id.asLong();
```
Comment (rouazana) [inline mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/json/IndexableMessage.java]: ternary?

File: mailbox/hbase/src/main/java/org/apache/james/mailbox/hbase/mail/HBaseMessageMapper.java [hunk 6]
```diff
@@ -189,12 +193,12 @@ public void endRequest() {
...
 
-    private List<MailboxMessage> findMessagesInMailboxAfterUID(HBaseId mailboxId, long from, int batchSize, boolean flaggedForDelete) throws IOException {
+    private List<MailboxMessage> findMessagesInMailboxAfterUID(HBaseId mailboxId, MessageUid messageUid, int batchSize, boolean flaggedForDelete) throws IOException {
         List<MailboxMessage> messageList = new ArrayList<MailboxMessage>();
         HTable messages = new HTable(conf, MESSAGES_TABLE);
         // uids are stored in reverse so we need to search
-        Scan scan = new Scan(messageRowKey(mailboxId, Long.MAX_VALUE),
-                messageRowKey(mailboxId, from - 1));
+        Scan scan = new Scan(messageRowKey(mailboxId, MessageUid.MAX_VALUE),
+                messageRowKey(mailboxId, messageUid.previous()));
```
Comment (rouazana) [inline mailbox/hbase/src/main/java/org/apache/james/mailbox/hbase/mail/HBaseMessageMapper.java]: should break when messageUid is 1, because it should call minMessageRowKey instead

File: mailbox/hbase/src/main/java/org/apache/james/mailbox/hbase/mail/HBaseMessageMapper.java [hunk 7]
```diff
@@ -219,16 +223,16 @@ public void endRequest() {
...
-        if (from > to) {
+        if (from.compareTo(to) > 0) {
             return messageList;
         }
         HTable messages = new HTable(conf, MESSAGES_TABLE);
         /*TODO: check if Between should be inclusive or exclusive regarding limits.
          * HBase scan operaion are exclusive to the upper bound when providing stop row key.
          */
-        Scan scan = new Scan(messageRowKey(mailboxId, to), messageRowKey(mailboxId, from - 1));
+        Scan scan = new Scan(messageRowKey(mailboxId, to), messageRowKey(mailboxId, from.previous()));
```
Comment (rouazana) [inline mailbox/hbase/src/main/java/org/apache/james/mailbox/hbase/mail/HBaseMessageMapper.java]: idem

File: mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/mail/model/openjpa/AbstractJPAMailboxMessage.java [hunk 8]
```diff
@@ -409,105 +391,76 @@ public Date getInternalDate() {
...
+         * (uFlags.contains(f.getName()) == false) { userFlags.remove(f); i++; }
+         * } for (int i = 0; i < uFlags.size(); i++) { boolean found = false;
+         * String uFlag = uFlags.get(i); for (int a = 0; a < userFlags.size();
+         * a++) { String userFlag = userFlags.get(a).getName(); if
+         * (userFlag.equals(uFlag)) { found = true; break; } } if (found ==
+         * false) { userFlags.add(new JPAUserFlag(uFlag)); }
+         * 
+         * 
+         * }
+         */
```
Comment (rouazana) [inline mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/mail/model/openjpa/AbstractJPAMailboxMessage.java]: eurk

File: mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/mail/model/openjpa/AbstractJPAMailboxMessage.java [hunk 9]
```diff
@@ -562,19 +515,12 @@ public int compareTo(MailboxMessage other) {
...
-        + "recent = " + this.recent + TOSTRING_SEPARATOR
-        + "seen = " + this.seen + TOSTRING_SEPARATOR
-        + " )";
-    }
-    
+        return "message(" + "mailboxId = " + this.getMailboxId() + TOSTRING_SEPARATOR + "uid = " + this.uid
+                + TOSTRING_SEPARATOR + "internalDate = " + this.internalDate + TOSTRING_SEPARATOR + "answered = "
+                + this.answered + TOSTRING_SEPARATOR + "deleted = " + this.deleted + TOSTRING_SEPARATOR + "draft = "
+                + this.draft + TOSTRING_SEPARATOR + "flagged = " + this.flagged + TOSTRING_SEPARATOR + "recent = "
+                + this.recent + TOSTRING_SEPARATOR + "seen = " + this.seen + TOSTRING_SEPARATOR + " )";
```
Comment (rouazana) [inline mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/mail/model/openjpa/AbstractJPAMailboxMessage.java]: previous formatting was better

File: mailbox/maildir/src/main/java/org/apache/james/mailbox/maildir/MaildirFolder.java [hunk 10]
```diff
@@ -181,20 +185,22 @@ public File getTmpFolder() {
...
     /**
      * Returns the last uid used in this mailbox
-     * @param session
-     * @return lastUid
-     * @throws MailboxException
      */
-    public long getLastUid(MailboxSession session) throws MailboxException {
-        if (lastUid == -1) {
+    public Optional<MessageUid> getLastUid(MailboxSession session) throws MailboxException {
+        if (lastUid == null) {
```
Comment (rouazana) [inline mailbox/maildir/src/main/java/org/apache/james/mailbox/maildir/MaildirFolder.java]: I know that's my code, but a null Optional is a bit disturbing, no?

File: mailbox/maildir/src/main/java/org/apache/james/mailbox/maildir/MaildirFolder.java [hunk 11]
```diff
@@ -765,7 +777,7 @@ public Long execute() throws MailboxException {
                         while ((line = reader.readLine()) != null)
                             lines.add(line);
                         uid = getNextUid();
-                        lines.add(String.valueOf(uid) + " " + name);
+                        lines.add(String.valueOf(uid.asLong()) + " " + name);
```
Comment (rouazana) [inline mailbox/maildir/src/main/java/org/apache/james/mailbox/maildir/MaildirFolder.java]: What about adding a MessageUid.asString?

File: mailbox/maildir/src/main/java/org/apache/james/mailbox/maildir/MaildirStore.java [hunk 12]
```diff
@@ -237,13 +241,15 @@ public String getFolderName(MailboxPath mailboxPath) {
...
 
-    /**
-     * @see org.apache.james.mailbox.store.mail.UidProvider#nextUid(org.apache.james.mailbox.MailboxSession, org.apache.james.mailbox.store.mail.model.Mailbox)
-     */
     @Override
-    public long nextUid(MailboxSession session, Mailbox mailbox) throws MailboxException {
+    public MessageUid nextUid(MailboxSession session, Mailbox mailbox) throws MailboxException {
         try {
-            return createMaildirFolder(mailbox).getLastUid(session) +1;
+            return createMaildirFolder(mailbox).getLastUid(session).transform(new Function<MessageUid, MessageUid>() {
```
Comment (rouazana) [inline mailbox/maildir/src/main/java/org/apache/james/mailbox/maildir/MaildirStore.java]: If all these functions would have been defined in MessageUid class, this would avoid these repetitions, no?

File: mailbox/zoo-seq-provider/src/test/java/org/apache/james/mailbox/store/mail/ZooUidProviderTest.java [hunk 13]
```diff
@@ -187,9 +189,9 @@ public void testLastUid() throws Exception {
 	    @Test
 	    public void testLongLastUid() throws Exception {
 	        System.out.println("Testing long lastUid");
-	        long result = longProvider.lastUid(null, mailboxLong);
-	        assertEquals("Next UID is 0", 0, result);
-	        result = longProvider.nextUid(null, mailboxLong);
-	        assertEquals("Next UID is 1", 1, result);
+	        Optional<MessageUid> result = longProvider.lastUid(null, mailboxLong);
+	        assertEquals("Next UID is 0", Optional.absent(), result);
```
Comment (rouazana) [inline mailbox/zoo-seq-provider/src/test/java/org/apache/james/mailbox/store/mail/ZooUidProviderTest.java]: bad description

File: mailbox/api/src/main/java/org/apache/james/mailbox/MessageUid.java (hunk 1 above)
Comment (aduprat) [inline mailbox/api/src/main/java/org/apache/james/mailbox/MessageUid.java]: Pitty it was not a Java 8 project:
`Long.compare(uid, o.uid)`

File: mailbox/api/src/main/java/org/apache/james/mailbox/model/MessageRange.java [hunk 14]
```diff
@@ -169,35 +157,35 @@ public String toString() {
...
         long firstUid = 0;
         int a = 0;
         for (int i = 0; i < uids.size(); i++) {
-            long u = uids.get(i);
+            long u = uids.get(i).asLong();
             if (i == 0) {
                 firstUid =  u;
                 if (uids.size() == 1) {
-                    ranges.add(MessageRange.one(firstUid));
+                    ranges.add(MessageRange.one(MessageUid.of(firstUid)));
```
Comment (aduprat) [inline mailbox/api/src/main/java/org/apache/james/mailbox/model/MessageRange.java]: `MessageUid.of(firstUid).toRange()`

File: mailbox/api/src/main/java/org/apache/james/mailbox/model/MessageRange.java [hunk 15]
```diff
@@ -169,35 +157,35 @@ public String toString() {
...
                 if ((firstUid + a +1) != u) {
-                    ranges.add(MessageRange.range(firstUid, firstUid + a));
+                    ranges.add(MessageRange.range(MessageUid.of(firstUid), MessageUid.of(firstUid + a)));
                     
                     // set the next first uid and reset the counter
                     firstUid = u;
                     a = 0;
                     if (uids.size() <= i +1) {
-                        ranges.add(MessageRange.one(firstUid));
+                        ranges.add(MessageRange.one(MessageUid.of(firstUid)));
```
Comment (aduprat) [inline mailbox/api/src/main/java/org/apache/james/mailbox/model/MessageRange.java]: same

File: mailbox/api/src/test/java/org/apache/james/mailbox/model/MessageRangeTest.java [hunk 16]
```diff
@@ -23,55 +23,133 @@
...
+                MessageRange.range(MessageUid.of(5), MessageUid.of(6)), 
+                MessageRange.one(MessageUid.of(9)));
     }
     
     @Test
     public void givenASingleNumberToRangeShouldReturnOneRange() {
-        List<MessageRange> ranges = MessageRange.toRanges(Arrays.asList(1L));
-        assertThat(ranges).containsExactly(MessageRange.one(1));
+        List<MessageRange> ranges = MessageRange.toRanges(Arrays.asList(MessageUid.of(1L)));
+        assertThat(ranges).containsExactly(MessageRange.one(MessageUid.of(1)));
```
Comment (aduprat) [inline mailbox/api/src/test/java/org/apache/james/mailbox/model/MessageRangeTest.java]: same

File: mailbox/caching/src/main/java/org/apache/james/mailbox/caching/MailboxMetadataCache.java [hunk 17]
```diff
@@ -17,10 +20,10 @@ long countMessagesInMailbox(Mailbox mailbox,
 	long countUnseenMessagesInMailbox(Mailbox mailbox,
 									  MessageMapper underlying) throws MailboxException;
```
Comment (aduprat) [inline mailbox/caching/src/main/java/org/apache/james/mailbox/caching/MailboxMetadataCache.java]: indent

File: mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/CassandraUidProviderTest.java [hunk 18]
```diff
@@ -67,24 +68,23 @@ public void cleanUp() {
     @Test
     public void lastUidShouldRetrieveValueStoredByNextUid() throws Exception {
         int nbEntries = 100;
-        long result = uidProvider.lastUid(null, mailbox);
-        assertEquals(0, result);
+        Optional<MessageUid> result = uidProvider.lastUid(null, mailbox);
+        assertThat(result).isEqualTo(Optional.absent());
```
Comment (aduprat) [inline mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/CassandraUidProviderTest.java]: assertj-guava

File: mailbox/hbase/src/test/java/org/apache/james/mailbox/hbase/mail/HBaseUidAndModSeqProviderTest.java [hunk 19]
```diff
@@ -129,11 +132,15 @@ public void testLastUid() throws Exception {
     public void testNextUid() throws Exception {
         LOG.info("nextUid");
         HBaseMailbox mailbox = mailboxList.get(mailboxList.size() / 2);
-        long lastUid = uidProvider.lastUid(null, mailbox);
-        long result;
-        for (int i = (int) lastUid + 1; i < (lastUid + 10); i++) {
-            result = uidProvider.nextUid(null, mailbox);
-            assertEquals(i, result);
+        Optional<MessageUid> lastUid = uidProvider.lastUid(null, mailbox);
```
Comment (aduprat) [inline mailbox/hbase/src/test/java/org/apache/james/mailbox/hbase/mail/HBaseUidAndModSeqProviderTest.java]: null ?

File: mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/CassandraUidProviderTest.java (hunk 4 above)
Comment (mbaechler) [inline mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/CassandraUidProviderTest.java]: it doesn't work because Function<A,B> requires A to be an object.

File: mailbox/api/src/main/java/org/apache/james/mailbox/MessageUid.java (hunk 1 above)
Comment (mbaechler) [inline mailbox/api/src/main/java/org/apache/james/mailbox/MessageUid.java]: `Long.valueOf(uid).compareTo(o.uid)` is just more ambigous because you compare two different types

File: mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/mail/model/openjpa/AbstractJPAMailboxMessage.java (hunk 8 above)
Comment (mbaechler) [inline mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/mail/model/openjpa/AbstractJPAMailboxMessage.java]: I have no idea why we have this change. I remove the commented code.

File: mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/mail/model/openjpa/AbstractJPAMailboxMessage.java (hunk 9 above)
Comment (mbaechler) [inline mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/mail/model/openjpa/AbstractJPAMailboxMessage.java]: of course, it looks like an automatic reformating

File: mailbox/maildir/src/main/java/org/apache/james/mailbox/maildir/MaildirFolder.java (hunk 11 above)
Comment (mbaechler) [inline mailbox/maildir/src/main/java/org/apache/james/mailbox/maildir/MaildirFolder.java]: I think it's not a good idea because uid is never defined as a string and you have no way to build on from a string.

File: mailbox/maildir/src/main/java/org/apache/james/mailbox/maildir/MaildirStore.java (hunk 12 above)
Comment (mbaechler) [inline mailbox/maildir/src/main/java/org/apache/james/mailbox/maildir/MaildirStore.java]: what, dealing with Optional ?

File: mailbox/maildir/src/main/java/org/apache/james/mailbox/maildir/MaildirFolder.java (hunk 11 above)
Comment (rouazana) [inline mailbox/maildir/src/main/java/org/apache/james/mailbox/maildir/MaildirFolder.java]: ok

File: mailbox/maildir/src/main/java/org/apache/james/mailbox/maildir/MaildirStore.java (hunk 12 above)
Comment (rouazana) [inline mailbox/maildir/src/main/java/org/apache/james/mailbox/maildir/MaildirStore.java]: that's the idea, but actually it will re-introduce guava dependency into MessageUid, so that's not feasible.

File: mailbox/api/src/test/java/org/apache/james/mailbox/model/MessageRangeTest.java [hunk 20]
```diff
@@ -23,55 +23,133 @@
...
+
+    @Test
+    public void includeShouldBeTrueWhenInRange() {
+        MessageRange range = MessageRange.range(MessageUid.of(3), MessageUid.of(6));
+        boolean actual = range.includes(MessageUid.of(4));
+        assertThat(actual).isTrue();
+    }
+
+    @Test
+    public void includeShouldBeTrueWhenEqualsToRange() {
```
Comment (chibenwa) [inline mailbox/api/src/test/java/org/apache/james/mailbox/model/MessageRangeTest.java]: You should also test inferior born of the range

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraUidProvider.java [hunk 21]
```diff
@@ -66,21 +66,20 @@ public CassandraUidProvider(Session session) {
     }
 
     @Override
-    public long nextUid(MailboxSession mailboxSession, Mailbox mailbox) throws MailboxException {
+    public MessageUid nextUid(MailboxSession mailboxSession, Mailbox mailbox) throws MailboxException {
         CassandraId mailboxId = (CassandraId) mailbox.getMailboxId();
-        if (findHighestUid(mailboxId).isFirst()) {
-            Optional<Uid> optional = tryInsertUid(mailboxId, FIRST_UID);
+        if (! findHighestUid(mailboxId).isPresent()) {
+            Optional<MessageUid> optional = tryInsertUid(mailboxId, com.google.common.base.Optional.absent());
```
Comment (chibenwa) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraUidProvider.java]: Why the use of Guava optionals here ? Isn't the use of Optional internal to the class ? I don't see the value added by the use of Guava optional....

File: mailbox/api/src/test/java/org/apache/james/mailbox/model/MessageRangeTest.java (hunk 20 above)
Comment (rouazana) [inline mailbox/api/src/test/java/org/apache/james/mailbox/model/MessageRangeTest.java]: that's just upper in includeShouldBeTrueWhenEqualsFromRange, no?

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageResultIterator.java [hunk 22]
```diff
@@ -253,7 +247,7 @@ public boolean equals(Object obj) {
             }
             if (obj instanceof UnloadedMessageResult) {
                 UnloadedMessageResult that = (UnloadedMessageResult)obj;
-                return (size == that.size) && (uid == that.uid) && (modSeq == that.modSeq) && exception.equals(that.exception)
+                return (size == that.size) && (uid.equals(that.uid)) && (modSeq == that.modSeq) && exception.equals(that.exception)
```
Comment (chibenwa) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageResultIterator.java]: Couldn't Guava's equals be used here ?

File: protocols/imap/src/main/java/org/apache/james/imap/api/message/UidRange.java [hunk 23]
```diff
@@ -0,0 +1,110 @@
...
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.imap.api.message;
```
Comment (chibenwa) [inline protocols/imap/src/main/java/org/apache/james/imap/api/message/UidRange.java]: Isn't this exactly what MessageRange is supposed to be ?

File: protocols/imap/src/main/java/org/apache/james/imap/api/message/response/StatusResponse.java [hunk 24]
```diff
@@ -152,6 +154,13 @@ public static ResponseCode condStore(IdRange[] failedRanges) {
             return new ResponseCode("MODIFIED", Arrays.asList(new String[] { failed}), 0, false);
         }
         
+        /** RFC4551 <code>Conditional STORE</code> response code */
```
Comment (chibenwa) [inline protocols/imap/src/main/java/org/apache/james/imap/api/message/response/StatusResponse.java]: It is more than a type change.... It looks like you correct conformity to the RFC here...
It is cool but it may be better not to do it in a 2500+ lines modified commit?

File: protocols/imap/src/main/java/org/apache/james/imap/encode/ESearchResponseEncoder.java [hunk 25]
```diff
@@ -66,10 +63,14 @@ protected void doEncode(ImapMessage acceptableMessage, ImapResponseComposer comp
         if (options.contains(SearchResultOption.COUNT)) {
             composer.message(SearchResultOption.COUNT.name()).message(count);
         }
-        if (all != null && all.length > 0 && options.contains(SearchResultOption.ALL)) {
+        if (!useUid && all != null && all.length > 0 && options.contains(SearchResultOption.ALL)) {
```
Comment (chibenwa) [inline protocols/imap/src/main/java/org/apache/james/imap/encode/ESearchResponseEncoder.java]: Again, it looks unrelated to a simple type change

File: protocols/imap/src/main/java/org/apache/james/imap/api/message/UidRange.java [hunk 26]
```diff
@@ -0,0 +1,110 @@
+/****************************************************************
```
Comment (chibenwa) [inline protocols/imap/src/main/java/org/apache/james/imap/api/message/UidRange.java]: No tests for this ?

File: protocols/imap/src/main/java/org/apache/james/imap/api/message/response/StatusResponse.java (hunk 24 above)
Comment (mbaechler) [inline protocols/imap/src/main/java/org/apache/james/imap/api/message/response/StatusResponse.java]: Not at all.
Previously, we used IdRange to store message sequence numbers or uids.
It's why you could live with a single condStore method. The new one is just a copy/paste of the previous that takes the new UidRange type. There's no logic change.

File: protocols/imap/src/main/java/org/apache/james/imap/encode/ESearchResponseEncoder.java (hunk 25 above)
Comment (mbaechler) [inline protocols/imap/src/main/java/org/apache/james/imap/encode/ESearchResponseEncoder.java]: I clearly don't agree.

File: mailbox/api/src/test/java/org/apache/james/mailbox/model/MessageRangeTest.java (hunk 20 above)
Comment (chibenwa) [inline mailbox/api/src/test/java/org/apache/james/mailbox/model/MessageRangeTest.java]: It is covered for FROM, but not for RANGE. Test battery should not expect code to be factorized between FROM and RANGE IMO

File: protocols/imap/src/main/java/org/apache/james/imap/encode/ESearchResponseEncoder.java (hunk 25 above)
Comment (chibenwa) [inline protocols/imap/src/main/java/org/apache/james/imap/encode/ESearchResponseEncoder.java]: I do not uderstand why a hudge part of the changes in imap subpart was needed. From what I see, UIDs can now be Optional, and there is also a type change from IdRange to UidRange... I found hard to review the IMAP changes, and it is not clear for me, why it is related to maibox's MessageUid....

Comment (mbaechler) [inline protocols/imap/src/main/java/org/apache/james/imap/encode/ESearchResponseEncoder.java]: I can understand that the changes are not easy to read but I tried to summarized in a previous answer : 
- in james core, we changed from Long to MessageUid for message uids.
- imap has the concept of "message sequence number" (msn). Before MessageUid, IdRange was used for the two types : uid and msn. Of course, we don't introduce MessageUid type to use such a poor concept so we introduced UidRange. IdRange is made of uids, IdRange is mande of msn. That's all you have to know.

Comment (chibenwa) [inline protocols/imap/src/main/java/org/apache/james/imap/encode/ESearchResponseEncoder.java]: Ok, thanks

File: mailbox/api/src/test/java/org/apache/james/mailbox/model/MessageRangeTest.java (hunk 20 above)
Comment (rouazana) [inline mailbox/api/src/test/java/org/apache/james/mailbox/model/MessageRangeTest.java]: I agree. But really includeShouldBeTrueWhenEqualsFromRange is dealing with RANGE, no?

Comment (chibenwa) [inline mailbox/api/src/test/java/org/apache/james/mailbox/model/MessageRangeTest.java]: I agree. It is. This test will not improve code coverage. But it will better explain specification, and make it obvious....

Comment (rouazana) [inline mailbox/api/src/test/java/org/apache/james/mailbox/model/MessageRangeTest.java]: I really think you missed my point.
We already have: includeShouldBeTrueWhenEqualsFrom (with a MessageRange.from) and includeShouldBeTrueWhenEqualsFromRange (with a MessageRange.range), so I don't see anything missing.

Comment (chibenwa) [inline mailbox/api/src/test/java/org/apache/james/mailbox/model/MessageRangeTest.java]: ok

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraUidProvider.java (hunk 21 above)
Comment (mbaechler) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraUidProvider.java]: good catch

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageResultIterator.java (hunk 22 above)
Comment (mbaechler) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageResultIterator.java]: It could, it's not what I decided to do, I preferred not refactor that piece of code.

Comment (chibenwa) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageResultIterator.java]: Ok

File: mailbox/api/src/test/java/org/apache/james/mailbox/MessageUidTest.java [hunk 27]
```diff
@@ -0,0 +1,39 @@
+package org.apache.james.mailbox;
```
Comment (aduprat) [inline mailbox/api/src/test/java/org/apache/james/mailbox/MessageUidTest.java]: license missing

File: mailbox/api/src/test/java/org/apache/james/mailbox/MessageUidTest.java [hunk 28]
```diff
@@ -0,0 +1,39 @@
...
+
+    @Test
+    public void distanceShouldReturnZeroWhenSameValue() {
+        assertThat(_1.distance(_1)).isEqualTo(0);
+    }
+
+    @Test
+    public void distanceShouldThrowWhenNullArgument() {
+        exception.expect(NullPointerException.class);
+        assertThat(_1.distance(null)).isEqualTo(0);
```
Comment (aduprat) [inline mailbox/api/src/test/java/org/apache/james/mailbox/MessageUidTest.java]: You don't need to assert

File: protocols/imap/src/test/java/org/apache/james/imap/api/message/UidRangeTest.java [hunk 29]
```diff
@@ -0,0 +1,154 @@
+package org.apache.james.imap.api.message;
```
Comment (aduprat) [inline protocols/imap/src/test/java/org/apache/james/imap/api/message/UidRangeTest.java]: license missing

File: protocols/imap/src/main/java/org/apache/james/imap/api/message/UidRange.java [hunk 30]
```diff
@@ -38,18 +38,56 @@
...
             rangeSet.add(Range.closed(range.getLowVal(), range.getHighVal()));
         }
-        List<UidRange> result = new ArrayList<UidRange>();
+        return rangeSet;
+    }
+
+    private static LinkedList<Range<MessageUid>> mergeContiguousRanges(RangeSet<MessageUid> rangeSet) {
+        LinkedList<Range<MessageUid>> mergedRanges = new LinkedList<Range<MessageUid>>();
+        
         for (Range<MessageUid> range: rangeSet.asRanges()) {
```
Comment (aduprat) [inline protocols/imap/src/main/java/org/apache/james/imap/api/message/UidRange.java]: Can you add a test with unordered ranges, to be sure that `TreeRangeSet` do the job:
[[1,2], [10,22], [3,4]] -> [[1,4], [10,22]]

**Comment by chibenwa** [comment]
test this please

**Comment by aduprat** [comment]
test this please

**Comment by aduprat** [comment]
test this please

**Comment by aduprat** [comment]
test this please

**Comment by aduprat** [comment]
test this please

**Comment by aduprat** [comment]
Will be merged just after squash
